javalib/src/main/scala/java/lang/dec2flt/FloatingPoint.scala

```diff
@@ -0,0 +1,152 @@
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package java.lang.dec2flt
+
+import java.math.BigInteger
+import java.lang.{Long => jlLong}
+
+/** DIY floating point number with 64-bit significand bits */
+private[dec2flt] class FloatingPoint private (val f: Long, val e: Int) {
+  import FloatingPoint._
+
+  override def toString(): String =
+    s"${this.f} * 2^${this.e}"
+
+  /** Returns a normalized product */
+  def mul(other: FloatingPoint): FloatingPoint = {
+    val low = f * other.f
+    val high = Math.unsignedMultiplyHigh(f, other.f)
+
+    val nlz = jlLong.numberOfLeadingZeros(high)
+    val excessBits = 64 - nlz
+    val shiftedF =
+      if (nlz == 0) high
+      else (high << nlz) | (low >>> excessBits)
+    val newE = (this.e + other.e + excessBits)
+
+    val remainder = low & (-1L >>> nlz)
+    val halfway = 1L << (excessBits - 1)
+
+    val newF = {
+      if (jlLong.compareUnsigned(remainder, halfway) > 0) // roundup
+        shiftedF + 1
+      else if (jlLong.compareUnsigned(remainder, halfway) < 0)
+        shiftedF
+      else if (shiftedF % 2 == 1) // tie, rounddown is odd
+        shiftedF + 1 // round to even
+      else
+        shiftedF
+    }
+    if (newF == 0) {
+      // denormalized by carry, in case shiftedF = 111..111
+      // (111...111 + 1) >> 1
+      new FloatingPoint(Long.MinValue, newE + 1)
+    } else {
+      new FloatingPoint(newF, newE)
+    }
+  }
+
+  def doesOverflowOnRound(fmt: FloatingPointFormat): Boolean = {
+    val (sig, exponent) = roundNormal(fmt)
+
+    // Remove the leading implicit bit
+    val encodedSig: Long = sig - (1 << (fmt.ExplicitSigBits))
+
+    // Adjust the exponent for exponent bias and mantissa shift
+    val encodedExp: Int = exponent + ((1 << (fmt.ExpBits - 1)) - 1) + fmt.ExplicitSigBits
+
+    encodedExp <= 0 || encodedExp >= (1 << fmt.ExpBits) ||
+        encodedSig <= 0 || encodedSig >= (1 << fmt.ExplicitSigBits)
+  }
+
+  /** Round the 64-bit significand to target#SigBits bits with half-to-even. */
+  private[dec2flt] def roundNormal(target: FloatingPointFormat): (Long, Int) = {
+    val excess = SigBits - target.SigBits // > 0
+    val halfway = 1L << (excess - 1)
+    val shiftedF = f >>> excess
+    val remainder = f & ((1L << excess) - 1) // > 0
+    val normalizedE = e + excess
+
+    val normalizedF = if (remainder > halfway) { // roundup
+      shiftedF + 1
+    } else if (remainder < halfway) { // rounddown
+      shiftedF
+    } else if (shiftedF % 2 == 1) { // tie, rownddown (shiftedF) is odd (1 at LSB)
+      shiftedF + 1 // roundup is even
+    } else { // tie, rounddown is even
+      shiftedF
+    }
+    if (normalizedF > target.MaxSig) {
+      (normalizedF >> 1, normalizedE + 1)
+    } else {
+      (normalizedF, normalizedE)
+    }
+  }
+}
+
+object FloatingPoint {
+  private val MASK: Long = 0xffffffffL
```

Comment:
Unfortunately this makes it not a constant in Scala. Write as
```suggestion
  private final val MASK = 0xffffffffL
```
(with `final` and without `: Long`)

(also name it `Mask` to follow Scala conventions?)

Reply:
Right, and it seems this is unused 😁 removed.